File: 2021/07-the-treatchery-of-whales/07.py
from collections import Counter

file = open("input.txt")
contents = file.read().strip()
file.close()

c = Counter(map(int, contents.split(",")))

def poschange(a, b):
    r = abs(a - b)
    return int((1 + r) / 2 * r)

res = []
for root in range(min(c), max(c) + 1):
    s = 0
    for position, count in c.items():
        s += count * poschange(root, position)
    res.append(s)

print(min(res))

File: 2021/07-the-treatchery-of-whales/07.hs
import Data.List.Split
import qualified Data.Map as Map
import Data.Map (Map)

main = readFile "test.txt" >>= print . solution . map read . splitOn ","

solution xs = minimum $ crabRave range crabs
    where crabs = combineCrabs xs
          range = crabRange $ Map.keys crabs

combineCrabs :: [Int] -> Map Int Int
combineCrabs = foldl (\map key -> Map.insertWith (+) key 1 map) Map.empty

crabRange crabs = [minimum(crabs)..maximum(crabs)]

crabRave range crabs = map (migrationCost crabs) range

migrationCost :: Map Int Int -> Int -> Int
migrationCost crabs to = sum . map (movingCost to) . Map.toList $ crabs

movingCost :: Int -> (Int, Int) -> Int
movingCost to (position, count) = count * (algSum $ abs $ position - to)

algSum x = (x + 1) * x `div` 2

